src/System.Net.Http.WinHttpHandler/src/System/Net/Http/WinHttpResponseParser.cs

```diff
@@ -130,104 +156,133 @@ public static uint GetResponseHeaderNumberInfo(SafeWinHttpHandle requestHandle,
+                IntPtr.Zero,
+                ref bufferLengthInBytes,
                 IntPtr.Zero))
             {
                 int lastError = Marshal.GetLastWin32Error();
-                if (lastError == Interop.WinHttp.ERROR_WINHTTP_HEADER_NOT_FOUND)
+
+                Debug.Assert(lastError != Interop.WinHttp.ERROR_WINHTTP_HEADER_NOT_FOUND);
+
+                if (lastError != Interop.WinHttp.ERROR_INSUFFICIENT_BUFFER)
                 {
-                    return null;
+                    throw WinHttpException.CreateExceptionUsingError(lastError);
                 }
+            }
 
-                if (lastError != Interop.WinHttp.ERROR_INSUFFICIENT_BUFFER)
+            return new char[bufferLengthInBytes / 2];
+        }
+
+        private static int GetResponseHeaderInfo(SafeWinHttpHandle requestHandle, uint infoLevel, char[] buffer)
+        {
+            Debug.Assert(buffer != null, "buffer must not be null.");
+            Debug.Assert(buffer.Length > 0, "buffer must not be empty.");
+
+            uint bufferLengthInBytes = (uint)buffer.Length * 2;
+
+            GCHandle pinnedHandle = GCHandle.Alloc(buffer, GCHandleType.Pinned);
+            try
+            {
+                if (!Interop.WinHttp.WinHttpQueryHeaders(
+                    requestHandle,
+                    infoLevel,
+                    Interop.WinHttp.WINHTTP_HEADER_NAME_BY_INDEX,
+                    pinnedHandle.AddrOfPinnedObject(),
+                    ref bufferLengthInBytes,
+                    IntPtr.Zero))
                 {
+                    int lastError = Marshal.GetLastWin32Error();
+
+                    if (lastError == Interop.WinHttp.ERROR_WINHTTP_HEADER_NOT_FOUND)
+                    {
+                        return 0;
+                    }
+
+                    Debug.Assert(lastError != Interop.WinHttp.ERROR_INSUFFICIENT_BUFFER, "buffer must be of sufficient size.");
+
                     throw WinHttpException.CreateExceptionUsingError(lastError);
                 }
             }
+            finally
+            {
+                pinnedHandle.Free();
+            }
 
-            // Allocate space for the buffer.
-            int charsNeeded = (int)bytesNeeded / 2;
-            var buffer = new StringBuilder(charsNeeded, charsNeeded);
+            return (int)bufferLengthInBytes / 2;
+        }
 
-            results = Interop.WinHttp.WinHttpQueryHeaders(
-                requestHandle,
-                infoLevel,
-                Interop.WinHttp.WINHTTP_HEADER_NAME_BY_INDEX,
-                buffer,
-                ref bytesNeeded,
-                IntPtr.Zero);
-            if (!results)
+        private static bool TryGetKnownReasonPhrase(HttpStatusCode statusCode, char[] array, int startIndex, int length, out string reasonPhrase)
+        {
+            CharArrayHelpers.DebugAssertArrayInputs(array, startIndex, length);
+
+            reasonPhrase = HttpStatusDescription.Get(statusCode);
```

Comment:
>Is this always correct? i.e. is it not possible for the server to send a different reason phrase than the canonical one associated with the status code?

This is NOT always correct.  The server can send any random string for the ReasonPhrase.  It is configurable on many servers and even the ASP.NET APIs allow for any string to be passed overriding the default one.

So, we always need to parse out the real reason phrase from what was received on the wire.